### ISSUE TITLE
RAdam with decoupled weight decay policy like AdamW

### ISSUE THREAD

**Issue by bilzard** [pytorch/pytorch#107282]
### 🚀 The feature, motivation and pitch

Currently, RAdam's weight-decay policy is same as Adam.
However, I believe some users want to use weight-decay policy same as AdamW.

How about implementing "RAdamW" optimizer, or, specifying weight decay policy as an optional argument?

c.f.) I've look through previous related issue:
https://github.com/pytorch/pytorch/issues/88836

### Alternatives

- implementing RAdamW optimizer
- modify RAdam optimizer and add feature to specify which weight decay policy is preferred (Adam/AdamW)

### Additional context

1. As written in the documentation, the official implementation uses AdamW-like weight decay policy.

https://github.com/LiyuanLucasLiu/RAdam/blob/master/radam/radam.py#L84

2. In the original paper, the author says they used AdamW-style decoupled weight decay policy.

> Following Loshchilov & Hutter (2018), we decouple weight decays in the
> vanilla Adam, Adam with warmup and RAdam in our experiments

https://arxiv.org/abs/1908.03265v4

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

**Comment by janeyx99** [comment]
We're happy to accept a similar change to what we did with NAdam (adding a decouple_weight_decay kwarg) for RAdam. Example: https://github.com/pytorch/pytorch/pull/103881

**Comment by bilzard** [comment]
It's just a random idea but, instead of designing numerous optimizers that combine various features, why not implement a primitive function as a single interface, allowing users to freely combine and utilize them according to their needs?
E.g. as Lookahead[1] is implemented as a decorator.

```python
optimizer = Adam(...)
optimizer = Lookahead(optimizer)
```

[1] https://github.com/alphadl/lookahead.pytorch

**Comment by janeyx99** [comment]
Yes, this is also something I've thought a little bit about, where we make the optimizers more modular and let people explore putting different pieces together. It would be interesting to see what components we could extract.

I'm guessing pure SGD would be the primitive, and certain pieces would involve:
- weight decay (and which type?) 
- momentum (which might be broken down more into something similar to the adaptive terms below)
- having an adaptive term multiplied to the `lr * grad` related to the exponential average
- having an adaptive term related to the averages squared, etc.
- whether or not to use the sign of the update
- ways to incorporate second orderness
- ...

In the end, these would become OptimizerOperators in a sense, like an intermediate between torch ops and our optimizer functions today. There are a few main things that would need to be clarified/unblocked before I'm convinced that this is the right way forward:
* Coverage: Is it possible to span an interesting enough group of optimizers with a closed set of features? For example, we would want to also be able to cover nonpointwise and second order optims such as LAMB.
* Abstraction: If fuller coverage requires introducing lots of ops, is it truly more useful than just using torch ops?
* Practicality: Will people want to invest in learning all the features to build their own optimizers? It sometimes feels like people just choose Adam(W) or SGD and roll with it. That said, maybe this is just a cycle that can be broken if a good enough offering existed.

As we ponder over these things, we shouldn't necessarily block RAdamW implementation if people desire it. 😛

**Comment by bilzard** [comment]
@janeyx99 The points are well-organized. Thank you for the feedback.

**Comment by bilzard** [comment]
@janeyx99 I made PR, but since it's first time to contribute to this repository, I don't know who to assign reviewer. Could you please assign reviewer?
https://github.com/pytorch/pytorch/pull/107507
